data-pipeline/src/agent_info/fetcher.rs

```diff
@@ -0,0 +1,349 @@
+        info: serde_json::from_reader(body_bytes.reader())?,
+    });
+    Ok(FetchInfoStatus::NewState(info))
+}
+
+/// Fetch the info endpoint once and return the info.
+///
+/// Can be used for one-time access to the agent's info. If you need to access the info several
+/// times use `AgentInfoFetcher` to keep the info up-to-date.
+///
+/// # Example
+/// ```no_run
+/// # use anyhow::Result;
+/// # #[tokio::main]
+/// # async fn main() -> Result<()> {
+/// // Define the endpoint
+/// let endpoint = ddcommon::Endpoint::from_url("http://localhost:8126/info".parse().unwrap());
+/// // Fetch the info
+/// let agent_info = data_pipeline::agent_info::fetch_info(&endpoint)
+///     .await
+///     .unwrap();
+/// println!("Agent version is {}", agent_info.info.version.unwrap());
+/// # Ok(())
+/// # }
+/// ```
+pub async fn fetch_info(info_endpoint: &Endpoint) -> Result<Box<AgentInfo>> {
+    match fetch_info_with_state(info_endpoint, None).await? {
+        FetchInfoStatus::NewState(info) => Ok(info),
+        FetchInfoStatus::SameState => Err(anyhow!("Invalid state header")),
+    }
+}
+
+/// Fetch the info endpoint and update an ArcSwap keeping it up-to-date.
+///
+/// Once the fetcher has been created you can get an Arc of the config by calling `get_info`.
+/// You can then start the run method, the fetcher will update the AgentInfoArc based on the
+/// given refresh interval.
+///
+/// # Example
+/// ```no_run
+/// # use anyhow::Result;
+/// # #[tokio::main]
+/// # async fn main() -> Result<()> {
+/// // Define the endpoint
+/// let endpoint = ddcommon::Endpoint::from_url("http://localhost:8126/info".parse().unwrap());
+/// // Create the fetcher
+/// let fetcher = data_pipeline::agent_info::AgentInfoFetcher::new(
+///     endpoint,
+///     std::time::Duration::from_secs(5 * 60),
+/// );
+/// // Get the Arc to access the info
+/// let agent_info_arc = fetcher.get_info();
+/// // Start the runner
+/// tokio::spawn(async move {
+///     fetcher.run().await;
+/// });
+///
+/// // Access the info
+/// if let Some(agent_info) = agent_info_arc.load().as_ref() {
+///     println!(
+///         "Agent version is {}",
+///         agent_info.info.version.as_ref().unwrap()
+///     );
+/// }
+/// # Ok(())
+/// # }
+/// ```
+pub struct AgentInfoFetcher {
+    info_endpoint: Endpoint,
+    info: AgentInfoArc,
+    refresh_interval: Duration,
+}
+
+impl AgentInfoFetcher {
+    /// Return a new `AgentInfoFetcher` fetching the `info_endpoint` on each `refresh_interval`
+    /// and updating the stored info.
+    pub fn new(info_endpoint: Endpoint, refresh_interval: Duration) -> Self {
+        Self {
+            info_endpoint,
+            info: Arc::new(ArcSwapOption::new(None)),
+            refresh_interval,
+        }
+    }
+
+    /// Start fetching the info endpoint with the given interval.
+    ///
+    /// # Warning
+    /// This method does not return and should be called within a dedicated task.
+    pub async fn run(&self) {
+        loop {
```

Comment:
Do we need some sort of graceful shutdown procedure here? What happens if the host tracer initiates a graceful shutdown?

Reply:
I not sure on how tokio will behave on shutdown. Since the host tracer created the runtime will it get shutdown has well ? The fetcher itself doesn't require any graceful shutdown as long it is shutdown as well.

Reply:
We don't need to address it as part of this PR, but we should probably create a ticket to confirm shutdown behavior for the data pipeline.